File: 2816/solution.java
class solution {
    int val;
    solution next;

    solution() {
    }

    solution(int val) {
        this.val = val;
    }

    solution(int val, solution next) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    public solution doubleIt(solution head) {
        head = reverseList(head);
        solution curr = head;
        int carry = 0;
        while (curr != null) {
            int total = (curr.val * 2 + carry) % 10;
            carry = (curr.val * 2 + carry) / 10;
            curr.val = total;
            curr = curr.next;
        }
        head = reverseList(head);
        if (carry != 0) {
            head = new solution(carry, head);
        }
        return head;
    }

    private solution reverseList(solution node) {
        solution prev = null;
        while (node != null) {
            solution next = node.next;
            node.next = prev;
            prev = node;
            node = next;
        }
        return prev;
    }
}

// time O(n)
// space O(1)

File: 2816/solution.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def doubleIt(self, head: Optional[ListNode]) -> Optional[ListNode]:
        def rev(node):
            curr, prev = node, None
            while curr:
                nxt = curr.next
                curr.next = prev
                prev, curr = curr, nxt
            return prev

        head = rev(head)
        cin = 0
        curr = head
        while curr:
            total = (curr.val * 2 + cin) % 10
            cin = (curr.val * 2 + cin) // 10
            curr.val = total
            curr = curr.next
        head = rev(head)
        if cin:
            head = ListNode(cin, head)
        return head
